jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PoolingPathResource.java

```diff
@@ -0,0 +1,271 @@
+        return uri == null ? null : Path.of(uri);
+    }
+
+    public static PoolingMount mount(URI uri) throws IOException
+    {
+        if (!uri.isAbsolute())
+            throw new IllegalArgumentException("not an absolute uri: " + uri);
+        if (PathResource.ALLOWED_SCHEMES.contains(uri.getScheme()))
+            throw new IllegalArgumentException("not an allowed scheme: " + uri);
+
+        try (AutoLock ignore = POOL_LOCK.lock())
+        {
+            FileSystem fileSystem;
+            try
+            {
+                fileSystem = FileSystems.newFileSystem(uri, Collections.emptyMap());
+            }
+            catch (FileSystemAlreadyExistsException fsaee)
+            {
+                fileSystem = Paths.get(uri).getFileSystem();
+            }
+            return retain(fileSystem, uri).mount;
+        }
+    }
+
+    public static Collection<PoolingMount> mounts()
+    {
+        try (AutoLock ignore = POOL_LOCK.lock())
+        {
+            return POOL.values().stream().map(m -> m.mount).toList();
+        }
+    }
+
+    public static void sweep()
+    {
+        Set<Map.Entry<FileSystem, Metadata>> entries;
+        try (AutoLock ignore = POOL_LOCK.lock())
+        {
+            entries = POOL.entrySet();
+        }
+
+        for (Map.Entry<FileSystem, Metadata> entry : entries)
+        {
+            FileSystem fileSystem = entry.getKey();
+            Metadata metadata = entry.getValue();
+
+            if (metadata.path == null)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Filesystem {} not backed by a file", fileSystem);
+                return;
+            }
+
+            try (AutoLock ignore = POOL_LOCK.lock())
+            {
+                // We must check if the FS is still open under the lock as a concurrent thread may have closed it.
+                if (fileSystem.isOpen() &&
+                    !Files.isReadable(metadata.path) ||
+                    !Files.getLastModifiedTime(metadata.path).equals(metadata.lastModifiedTime) ||
+                    Files.size(metadata.path) != metadata.size)
+                {
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("File {} backing filesystem {} has been removed or changed, closing it", metadata.path, fileSystem);
+                    POOL.remove(fileSystem);
+                    IO.close(fileSystem);
+                }
+            }
+            catch (IOException e)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Cannot read last access time or size of file {} backing filesystem {}", metadata.path, fileSystem);
+            }
+        }
+    }
+
+    private static URI containerUri(URI uri)
+    {
+        String scheme = uri.getScheme();
+        if ((scheme == null) || !scheme.equalsIgnoreCase("jar"))
+            return null;
+
+        String spec = uri.getRawSchemeSpecificPart();
+        int sep = spec.indexOf("!/");
```

Comment:
"!/" is something that applies only to a `JarPathResource` and not to a general `PoolingPathResource`

Reply:
which is why I added a scheme check for `jar`. But I agree that this method feels out of place, I'm just unsure what could be done to make this feel less awkward, at least for now.

Reply:
Create a `JarPathResource` class and put it there, along with some optimised methods.